lib/testobject.js

```diff
@@ -0,0 +1,164 @@
+import zip from './s3';
+// import { exec } from 'teen_process';
+import { fs } from 'appium-support';
+import log from '../lib/logger';
+import request from 'request-promise';
+import _ from 'lodash';
+
+const TestObject = {};
+
+TestObject.DEFAULT_DEVICE = 'HTC_One_M8_real';
+TestObject.DEFAULT_IOS_DEVICE = 'iPad_2_16GB_real';
+
+/**
+ * Takes desired capabilities and extends it with testobject capabilities
+ */
+TestObject.getTestObjectCaps = async function (caps = {}) {
+  // Upload the app to TestObject
+  let testobject_app_id;
+  try {
+    testobject_app_id = await TestObject.uploadTestObjectApp(caps.app);
+  } catch (e) {
+    throw new Error(`Could not upload ${caps.app} to TestObject: ${e.message}`);
+  }
+
+  // Determine the TestObject_device
+  let testobject_device;
+  if (process.env.TESTOBJECT_DEVICE) {
+    testobject_device = process.env.TESTOBJECT_DEVICE;
+  } else if (caps.platformName && caps.platformName.toLowerCase() === 'ios') {
+    testobject_device = TestObject.DEFAULT_IOS_DEVICE;
+  } else {
+    testobject_device = TestObject.DEFAULT_DEVICE;
+  }
+
+  // Extend the provided caps with special testobject caps 
+  // (see https://app.staging.testobject.org/#/danielgraham/contact-manager/appium/basic/instructions)
+  return  Object.assign(caps, {
+    testobject_app_id,
+    testobject_device,
+    testobject_appium_version: '1.6.0',
+    testobject_api_key: process.env.TESTOBJECT_API_KEY,
+    testobject_cache_device: true,
+  });
+};
+
+TestObject._appIdCache = {};
+
+// Make request-promise mockable
+TestObject.request = request;
+
+/**
+ * Uploads your app (.app, .apk, .ipa etc...) to TestObject server
+ */
+TestObject.uploadTestObjectApp = async function (app) {
+  if (!process.env.TESTOBJECT_API_KEY) {
+    throw new Error('Must provide TESTOBJECT_API_KEY in environment variables');
+  }
+  if (!process.env.TESTOBJECT_USERNAME) {
+    throw new Error('Must define TESTOBJECT_USERNAME in environment variables');
+  }
+
+  // If this app was already uploaded and hasn't changed since last upload, re-use the app id
+  const cache = TestObject._appIdCache[app];
+  if (cache) {
+    const {mtime} = await fs.stat(app);
+    const modifiedTime = +(new Date(mtime));
+    if (modifiedTime < cache.uploaded) {
+      return cache.id;
+    }
+  }
+
+  // Upload the app
+  // (see https://app.staging.testobject.org/#/danielgraham/contact-manager/appium/basic/instructions)
+  // TODO: Blocked on TestObject fixing this
+  /*const appId = await exec('curl', [
+    '-u', `"${process.env.TESTOBJECT_USERNAME}:${process.env.TESTOBJECT_API_KEY}"`,
+    '-X', 'POST', 'https://app.staging.testobject.org:443/api/storage/upload',
+    '-H', '"Content-Type: application/octet-stream"',
+    '---data-binary', `@${app}`,
+  ]);*/
+  const appId = null;
+
+  TestObject._appIdCache[app] = {
+    id: appId,
+    uploaded: +(new Date()),
+  };
+
+  return appId;
+};
+
+TestObject.MOCHA_TESTOBJECT_TIMEOUT = 10 * 60 * 1000; // Ten minutes
```

Comment:
Are these constants accessed outside this file? If not, they don't need to be on the object, and can be defined as `const`s at the top of the file.